common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+
+    std::shared_ptr<Socket> socket() override { return m_socket; }
+    int receive(Buffer &buffer) override
+    {
+        return ::recv(m_socket->fd(), buffer.data(), buffer.size(), MSG_DONTWAIT);
+    }
+    int receive_from(Buffer &buffer, SockAddr &address) override
+    {
+        return ::recvfrom(m_socket->fd(), buffer.data(), buffer.size(), MSG_DONTWAIT,
+                          address.sockaddr(), &address.length());
+    }
+    int receive_message(MsgHdr &message) override
+    {
+        // TODO
+        return -1;
+    }
+    int send(const Buffer &buffer, size_t length) override
+    {
+        return ::send(m_socket->fd(), buffer.data(), length, MSG_NOSIGNAL);
+    }
+    int send_to(const Buffer &buffer, size_t length, const SockAddr &address) override
+    {
+        return ::sendto(m_socket->fd(), buffer.data(), buffer.size(), 0, address.sockaddr(),
+                        address.length());
+    }
+    int send_message(const MsgHdr &message) override
+    {
+        // TODO
+        return -1;
+    }
+
+private:
+    /**
+     * Connected socket used by this connection object.
+     */
+    std::shared_ptr<Socket> m_socket;
+};
+
+class ServerSocketAbstractImpl : public ServerSocket {
+public:
+    int bind(const SockAddr &address)
+    {
+        return ::bind(m_socket->fd(), address.sockaddr(), address.length());
+    }
+
+    int listen(int backlog) { return ::listen(m_socket->fd(), backlog); }
+
+    std::unique_ptr<Connection> accept(SockAddr &address) override
+    {
+        int s = ::accept(m_socket->fd(), address.sockaddr(), &address.length());
+        if (FileDescriptor::invalid_descriptor == s) {
+            return nullptr;
+        }
+
+        return std::make_unique<SocketConnectionImpl>(std::make_shared<ConnectedSocket>(s));
+    }
+
+protected:
+    explicit ServerSocketAbstractImpl(const std::shared_ptr<Socket> &socket) : m_socket(socket) {}
+    std::shared_ptr<Socket> m_socket;
+};
+
+class ClientSocketAbstractImpl : public ClientSocket {
+public:
+    // TODO: add connection timeout
+    std::unique_ptr<Connection> connect(const SockAddr &address) override
+    {
+        if (0 != ::connect(m_socket->fd(), address.sockaddr(), address.length())) {
+            return nullptr;
+        }
+        return std::make_unique<SocketConnectionImpl>(m_socket);
+    }
+
+protected:
+    explicit ClientSocketAbstractImpl(const std::shared_ptr<Socket> &socket) : m_socket(socket) {}
+    std::shared_ptr<Socket> m_socket;
+};
+
+class RawServerSocket : public ServerSocketAbstractImpl {
+public:
+    explicit RawServerSocket(const std::shared_ptr<RawSocket> &socket)
+        : ServerSocketAbstractImpl(socket)
+    {
+    }
+};
+
+class RawClientSocket : public ClientSocketAbstractImpl {
+public:
+    explicit RawClientSocket(const std::shared_ptr<RawSocket> &socket)
+        : ClientSocketAbstractImpl(socket)
+    {
+    }
+};

```

Comment:
Actually, this one doesn't make sense for raw sockets either - there is no "connect".